design/design-doc-rbac-manager.md

```diff
@@ -0,0 +1,843 @@
+hand must be created by the RBAC manager, though their base roles can be created
+outside of Crossplane.
+
+The naming convention established by this design is that user-facing roles use
+hyphen separated names, while system facing and aggregated roles use colon
+user-facing role. There is only one `crossplane:aggregate-to-admin` base role
+and only separated names. Note that aggregation may be transitive. For example
+the `crossplane-admin` role is identical to the `crossplane-edit` role except
+for a few extra rules in its base role. Provider and composite resource roles
+can therefore aggregate only to `crossplane-edit`, which in turn aggregates to
+`crossplane-admin`.
+
+### Composite Resource ClusterRole Mechanics
+
+When a composite resource is defined the RBAC manager will create two cluster
+roles. For example creating an XRD named `composites.example.org` would trigger
+the creation of the following roles:
+
+* `crossplane:composite:composites.example.org:aggregate-to-edit`
+* `crossplane:composite:composites.example.org:aggregate-to-view`
+
+The `aggregate-to-edit` role aggregates to the `crossplane`, `crossplane-admin`,
+and `crossplane-edit` cluster roles, as well as any namespaced `edit` role for a
+namespace in which the composite resource may claimed. The `aggregate-to-view`
+cluster role aggregates to the `crossplane-view` cluster role, and any `view`
+role for a namespace in which the composite resource may be claimed. Below is an
+example of an `edit` composite resource cluster role.
+
+```yaml
+# This ClusterRole aggregates up to any namespaced role that declares
+# compatibility by including the XRD label. The same role also aggregates to the
+# crossplane-edit role (and thus transitively to the crossplane-admin role).
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: crossplane:composite:examplecomposites.xr.example.org:aggregate-to-edit
+  labels:
+    rbac.crossplane.io/aggregate-to-edit: "true"
+    rbac.crossplane.io/aggregate-to-ns-edit: "true"
+    rbac.crossplane.io/xrd: examplecomposites.xr.example.org
+rules:
+- apiGroups: [xr.example.org]
+  resources:
+  - exampleclaims
+  - exampleclaims/status
+  # Composite resources are cluster scoped. Binding this cluster role at the
+  # namespace scope will not actually grant access to these composite resources.
+  # Including them in this cluster role allows one crossplane:composite role to
+  # aggregate to both the crossplane-ns-example-edit and crossplane-edit roles.
+  - examplecomposites
+  - examplecomposites/status
+  verbs: ["*"]
+```
+
+Composite resources may be claimed in any namespace that has an annotation with
+the key `rbac.crossplane.io/composites.example.org`, set to `xrd-claim-accepted`
+(where `composites.example.org` is the name of the XRD). When the RBAC manager
+encounters a namespace with one or more such annotations, it creates `admin`,
+`edit` and `view` roles in that namespace.
```

Comment:
Suggestion: If one happens to uninstall or upgrade packages these annotations might become out of date. As well, if there's a typo, it wouldn't be obvious that it might fail to provision permissions for a type, or that it was an out of date permission grant.

For someone building a UI around this like Upbound cloud, listing a particular Kind, and seeing the status of whether the permission could be applied, or if it's deprecated/no longer existing would be a lot easier to work with, than a totally different component reading the namespace annotations, parsing the xrd out of them, and comparing these names with what's installed on the server to know what's still valid.

Reply:
@lukeweber Could you sketch me some YAML representing the resource you're thinking of?

Reply:
The only point was just that the calculation of whether you were able to find such an xrd will technically already be known by the RBAC controller and might make for a better DX, but as you noted it's another type. I'll defer to your judgement here. Just a suggestion :)

```yaml
apiVersion: rbac.authorization.crossplane.io/v1
kind: XRDGrant
metadata:
  name: composites.example.org
  namespace: foo
spec:
  xrd: composites.example.org
status:
  phases: ready , failed
  conditions: xrdMissing, createdRBAC, noPermissions
```

Reply:
Got it. It seems like the key value here is that we could use the new type's status to reflect whether the intent to accept/grant a claim in a namespace was successful. I'd like to stick with the current design for now, but we can introduce this type if annotations prove to be a poor UX. I believe it's fairly common in Kubernetes for something to start as an annotation and later grow into a real field/type.